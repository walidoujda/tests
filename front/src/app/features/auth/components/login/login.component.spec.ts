import { ComponentFixture, TestBed } from '@angular/core/testing';
import { FormBuilder, FormsModule, ReactiveFormsModule } from '@angular/forms';
import { RouterTestingModule } from '@angular/router/testing';
import { LoginComponent } from './login.component';
import { AuthService } from '../../services/auth.service';
import { Router } from '@angular/router';
import { of, throwError } from 'rxjs';
import { expect } from '@jest/globals';
import { HttpClientTestingModule } from '@angular/common/http/testing';
import { SessionService } from 'src/app/services/session.service';

describe('LoginComponent', () => {
  let component: LoginComponent;
  let fixture: ComponentFixture<LoginComponent>;

  beforeEach(() => {
    TestBed.configureTestingModule({
      declarations: [LoginComponent],
      imports: [ReactiveFormsModule, FormsModule, RouterTestingModule, HttpClientTestingModule],
      providers: [AuthService, FormBuilder, SessionService],
    });

    fixture = TestBed.createComponent(LoginComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should call authService.login and navigate to /sessions on successful login', () => {
    const authService = TestBed.inject(AuthService);
    const sessionService = TestBed.inject(SessionService);
    const router = TestBed.inject(Router);
    const loginRequest = { email: 'test@example.com', password: 'password' };
    const sessionInformation = {   token: 'token',
      type: 'type',
      id: 0,
      username: 'username',
      firstName: 'firstName',
      lastName: 'lastName',
      admin: true
    };

    jest.spyOn(authService, 'login').mockImplementation(() => of(sessionInformation));
    jest.spyOn(sessionService, 'logIn');
    jest.spyOn(router, 'navigate');

    component.form.setValue(loginRequest);
    component.submit();

    expect(authService.login).toHaveBeenCalledWith(loginRequest);
    expect(sessionService.logIn).toHaveBeenCalledWith(sessionInformation);
    expect(router.navigate).toHaveBeenCalledWith(['/sessions']);
  });

  it('should set onError to true on login error', () => {
    const authService = TestBed.inject(AuthService);
    const error = new Error('Login error');

    jest.spyOn(authService, 'login').mockImplementation(() => throwError(() => new Error('Login error')));

    component.submit();

    expect(component.onError).toBeTruthy();
  });

  it('should match snapshot', () => {
    expect(fixture).toMatchSnapshot();
  });

});
