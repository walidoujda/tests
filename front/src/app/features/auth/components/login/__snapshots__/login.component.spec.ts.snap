// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginComponent should match snapshot 1`] = `
<app-login
  __ngContext__={[Function Number]}
  authService={[Function AuthService]}
  fb={[Function FormBuilder]}
  form={[Function FormGroup]}
  hide={[Function Boolean]}
  onError="false"
  router={[Function Router]}
  sessionService={[Function SessionService]}
>
  <div
    class="login"
    fxlayout="row"
    fxlayoutalign="center center"
  >
    <mat-card>
      <mat-card-header
        fxlayoutalign="center center"
      >
        <mat-card-title>
          Login
        </mat-card-title>
      </mat-card-header>
      <form
        class="login-form ng-untouched ng-pristine ng-invalid"
        novalidate=""
      >
        <mat-card-content
          fxlayout="column"
        >
          <mat-form-field
            fxflex=""
          >
            <input
              class="ng-untouched ng-pristine ng-invalid"
              formcontrolname="email"
              matinput=""
              placeholder="Email"
            />
          </mat-form-field>
          <mat-form-field
            fxflex=""
          >
            <input
              class="ng-untouched ng-pristine ng-invalid"
              formcontrolname="password"
              matinput=""
              placeholder="Password"
              type="password"
            />
            <button
              aria-label="Hide password"
              aria-pressed="true"
              mat-icon-button=""
              matsuffix=""
            >
              <mat-icon>
                visibility_off
              </mat-icon>
            </button>
          </mat-form-field>
        </mat-card-content>
        <button
          color="primary"
          disabled=""
          mat-raised-button=""
          type="submit"
        >
          Submit
        </button>
      </form>
    </mat-card>
  </div>
</app-login>
`;
